### PR TITLE
`mongo` command.

### PR DESCRIPTION
**Pull request by BreadcrumbIsTaken**
### `mongo` Command

Adds support from [Issue #2350](https://github.com/DenizenScript/Denizen/issues/2350) on the DenizenScript/Denizen repo.

#### Things added
- `mongo` command
    - Provides actions:
        - connect -- Connect to MongoDB
        - disconnect -- Disconnect from MongoDB
        - command -- Run a simple command.
        - ~~command_map -- Run a complex command with multiple parameters and operations. Used for commands not wrapped by `mongo`.~~ Functionality now apart of the `command` action above.
        - find -- Find a specific Document using provided information. 
        - insert -- Insert a Document into current Collection and Database
        - update -- Update a specific Document using a query and new data.
        - use_database -- Change current Database being connected to.
        - use_collection -- Change current Collection being connected to.
- Tag `<util.mongo_connections>` -- Get the IDs of current Mongo connections in a ListTag.

Yes this is a very large PR, just bare with me 😅.
Here is a list of tasks that can be used to test the command: https://paste.denizenscript.com/View/101730
Uses test data from [here](https://www.mongodb.com/docs/atlas/sample-data/)

#### Notes
- MongoDB's Java Driver spits a whole bunch of stuff into the console when you create a new connections, almost to a point where it spams. It includes things like the client, clusters, and other various information.
- ~~I am unsure if the new `autoExecute` way of doing commands works for `mongo`. If it does I'll be happy to make a commit to update `parseArgs` and `execute` to the new way of doing things.~~ Now updated to use `autoExecute`.

###### 🌿🌿🌿

### CODE REVIEW

File: src/main/java/com/denizenscript/denizencore/scripts/commands/core/MongoCommand.java [hunk 1]
```diff
@@ -0,0 +1,756 @@
...
+    public static Map<String, MongoClient> connections = new HashMap<>();
+    public static Map<String, MongoDatabase> databases = new HashMap<>();
+    public static Map<String, MongoCollection<Document>> collections = new HashMap<>();
+
+    @Override
+    public void onDisable() {
+        for (Map.Entry<String, MongoClient> entry : connections.entrySet()) {
+            try {
+                entry.getValue().close();
+            } catch (final Exception e) {
```
Comment (tal5) [inline src/main/java/com/denizenscript/denizencore/scripts/commands/core/MongoCommand.java:151]: Java formatting - `catch` should be on a new line

**Comment by tal5** [comment]
> I am unsure if the new autoExecute way of doing commands works for mongo. If it does I'll be happy to make a commit to update parseArgs and execute to the new way of doing things.

It's not really related to what the command does (in most cases at least, some commands have more complex arguments which can't be updated yet iirc), just a new cleaner way of handling command arguments - so yeah should use the modern way

File: src/main/java/com/denizenscript/denizencore/scripts/commands/core/MongoCommand.java [hunk 2]
```diff
@@ -0,0 +1,756 @@
...
+                        }, 0));
+                    }
+                    if (CoreConfiguration.debugVerbose) {
+                        Debug.echoDebug(scriptEntry, "Connection did not error.");
+                    }
+                    final MongoClient conn = con;
+                    final MongoDatabase connDB = db;
+                    final MongoCollection<Document> coll = col;
+                    if (con != null) {
+                        DenizenCore.schedule(new OneTimeSchedulable(() -> {
```
Comment (tal5) [inline src/main/java/com/denizenscript/denizencore/scripts/commands/core/MongoCommand.java:336]: `DenizenCore#runOnMainThread`

File: src/main/java/com/denizenscript/denizencore/scripts/commands/core/MongoCommand.java [hunk 3]
```diff
@@ -0,0 +1,756 @@
...
+                            connections.put(id.asLowerString(), conn);
+                            databases.put(id.asLowerString(), connDB);
+                            collections.put(id.asLowerString(), coll);
+                            Debug.echoDebug(scriptEntry, "Successfully connected to Mongo server.");
+                            scriptEntry.setFinished(true);
+                        }, 0));
+                    } else {
```
Comment (tal5) [inline src/main/java/com/denizenscript/denizencore/scripts/commands/core/MongoCommand.java:343]: Java formatting - `else` should be on a new line

File: src/main/java/com/denizenscript/denizencore/tags/core/UtilTagBase.java [hunk 4]
```diff
@@ -952,6 +953,20 @@ else if (attribute.startsWith("format", 2) && attribute.hasParam()) {
         tagProcessor.registerTag(ElementTag.class, "stack_trace", (attribute, object) -> {
             return new ElementTag(DebugInternals.getFullExceptionMessage(new RuntimeException("TRACE"), false));
         });
+
+        // <--[tag]
+        // @attribute <util.mongo_connections>
+        // @returns ElementTag
+        // @description
+        // Returns a list of all Mongo connections opened by <@link command mongo>
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/denizencore/tags/core/UtilTagBase.java:961]: period at end.
Also add this tag next to the matching sql/redis tags so they're together.

File: src/main/java/com/denizenscript/denizencore/scripts/commands/core/MongoCommand.java [hunk 5]
```diff
@@ -0,0 +1,765 @@
...
+        if (!scriptEntry.hasObject("action")) {
+            throw new InvalidArgumentsException("Must specify an action!");
+        }
+        if (!scriptEntry.hasObject("upsert")) {
+            scriptEntry.defaultObject("upsert", new ElementTag("false"));
+        }
+    }
+
+    @Override
+    public void execute(final ScriptEntry scriptEntry) {
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/denizencore/scripts/commands/core/MongoCommand.java:257]: Use the new autoExecute, don't PR new commands using old syntaxes (this is actually two entire command registration formats behind here - the newer arg handler was the `argAsX` system, and autoExecute is newest)

File: src/main/java/com/denizenscript/denizencore/scripts/commands/core/MongoCommand.java [hunk 6]
```diff
@@ -0,0 +1,765 @@
...
+import org.bson.types.ObjectId;
+
+import java.util.*;
+
+public class MongoCommand extends AbstractCommand implements Holdable {
+    public MongoCommand() {
+        setName("mongo");
+        setSyntax("mongo [id:<ID>] [connect:<uri> database:<database> collection:<collection>/disconnect/command:<command> parameter:<parameter>/command_map:<map>/find:<map> (by_id:<id>)/insert:<map>/update:<update> new:<new> (upsert:true/{false})/use_database:<database>/use_collection:<collection>");
+        setRequiredArguments(2, 4);
+        allowedDynamicPrefixes = true;
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/denizencore/scripts/commands/core/MongoCommand.java:34]: Why?

File: src/main/java/com/denizenscript/denizencore/scripts/commands/core/MongoCommand.java [hunk 7]
```diff
@@ -0,0 +1,765 @@
...
+        isProcedural = false;
+    }
+
+    // <--[command]
+    // @Name Mongo
+    // @Short Interacts with a MongoDB server.
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/denizencore/scripts/commands/core/MongoCommand.java:40]: syntax info is missing

File: src/main/java/com/denizenscript/denizencore/scripts/commands/core/MongoCommand.java [hunk 8]
```diff
@@ -0,0 +1,765 @@
...
+import com.mongodb.client.result.UpdateResult;
+import org.bson.*;
+import org.bson.types.ObjectId;
+
+import java.util.*;
+
+public class MongoCommand extends AbstractCommand implements Holdable {
+    public MongoCommand() {
+        setName("mongo");
+        setSyntax("mongo [id:<ID>] [connect:<uri> database:<database> collection:<collection>/disconnect/command:<command> parameter:<parameter>/command_map:<map>/find:<map> (by_id:<id>)/insert:<map>/update:<update> new:<new> (upsert:true/{false})/use_database:<database>/use_collection:<collection>");
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/denizencore/scripts/commands/core/MongoCommand.java:32]: This appears to be misformatted? Not sure what all is going on here, but among other things, there's a `[` at the start with no end

File: src/main/java/com/denizenscript/denizencore/scripts/commands/core/MongoCommand.java [hunk 9]
```diff
@@ -0,0 +1,765 @@
...
+    // @Description
+    // This command is used to interact with a MongoDB server.
+    //
+    // MongoDB is a NoSQL database which uses concepts such as Documents and Collections to store data. MongoDB uses a form of JSON to represent its data.
+    // It can interact with localhost connections as well as hosted connections (such as MongoDB's Atlas) via a connection URI.
+    // Store your connection URI in the Denizen secrets file at 'plugins/Denizen/secrets.secret'. Refer to <@link ObjectType SecretTag> for usage info.
+    //
+    // Mongo works as a document-oriented database, where data is stored in Documents. Documents are stored inside Collections. Collections can contain many Documents. Collections are then stored inside Databases.
+    //
+    // Usage of the mongo command should almost always be used as ~awaitable (see <@link language ~waitable>), as large queries and insertions can take a while to retrieve or send.
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/denizencore/scripts/commands/core/MongoCommand.java:51]: typo on `waitable`

File: src/main/java/com/denizenscript/denizencore/scripts/commands/core/MongoCommand.java [hunk 10]
```diff
@@ -0,0 +1,765 @@
...
+import com.mongodb.client.model.UpdateOptions;
+import com.mongodb.client.result.InsertOneResult;
+import com.mongodb.client.result.UpdateResult;
+import org.bson.*;
+import org.bson.types.ObjectId;
+
+import java.util.*;
+
+public class MongoCommand extends AbstractCommand implements Holdable {
+    public MongoCommand() {
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/denizencore/scripts/commands/core/MongoCommand.java:30]: missing newline above

File: src/main/java/com/denizenscript/denizencore/scripts/commands/core/MongoCommand.java (hunk 5 above)
Comment (mcmonkey4eva) [reply src/main/java/com/denizenscript/denizencore/scripts/commands/core/MongoCommand.java:257]: Also, need a CoreConfiguration option to disable mongo - should be disabled by default even
Same as redis/sql have

File: src/main/java/com/denizenscript/denizencore/scripts/commands/core/MongoCommand.java [hunk 11]
```diff
@@ -0,0 +1,765 @@
...
+        }
+        try {
+            if (action.asString().equalsIgnoreCase("connect")) {
+                if (connections.containsKey(id.asString().toLowerCase())) {
+                    Debug.echoError(scriptEntry, "Already connected to a server with ID '" + id.asString() + "'!");
+                    scriptEntry.setFinished(true);
+                    return;
+                }
+                String rawUri;
+                if (uri.shouldBeType(SecretTag.class)) {
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/denizencore/scripts/commands/core/MongoCommand.java:285]: You copy/pasted SecretTag references that are written weird because they have backwards-compatibility with old formats - just require a SecretTag directly, don't use this conversion logic

File: src/main/java/com/denizenscript/denizencore/scripts/commands/core/MongoCommand.java [hunk 12]
```diff
@@ -0,0 +1,765 @@
...
+                    return;
+                }
+                DenizenCore.schedule(new AsyncSchedulable(new OneTimeSchedulable(() -> {
+                    MongoClient con = null;
+                    String conStr = rawUri;
+                    if (!conStr.startsWith("mongodb://") && !conStr.startsWith("mongodb+srv://")) {
+                        conStr = "mongodb://" + conStr;
+                    }
+                    if (CoreConfiguration.debugVerbose) {
+                        Debug.echoDebug(scriptEntry, "Connecting to Mongo server. . .");
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/denizencore/scripts/commands/core/MongoCommand.java:314]: Minor note: `...` no spaces

File: src/main/java/com/denizenscript/denizencore/scripts/commands/core/MongoCommand.java [hunk 13]
```diff
@@ -0,0 +1,765 @@
...
+                            databases.put(id.asLowerString(), connDB);
+                            collections.put(id.asLowerString(), coll);
+                            Debug.echoDebug(scriptEntry, "Successfully connected to Mongo server.");
+                            scriptEntry.setFinished(true);
+                        });
+                    }
+                    else {
+                        DenizenCore.runOnMainThread(() -> {
+                            scriptEntry.setFinished(true);
+                            if (CoreConfiguration.debugVerbose) {
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/denizencore/scripts/commands/core/MongoCommand.java:350]: Why the verbose check?

**Comment by BreadcrumbIsTaken** [comment]
Here are updated test scripts using the new prefixes and also another test that I wanted to add: https://paste.denizenscript.com/View/101730

File: src/main/java/com/denizenscript/denizencore/scripts/commands/core/MongoCommand.java [hunk 14]
```diff
@@ -0,0 +1,547 @@
...
+            return;
+        }
+        if (id == null) {
+            Debug.echoError(scriptEntry, "Missing connection ID!");
+            return;
+        }
+        Runnable runnable = () -> {
+            try {
+                if (uri != null) {
+                    if (connections.containsKey(id.toLowerCase())) {
```
Comment (tal5) [inline src/main/java/com/denizenscript/denizencore/scripts/commands/core/MongoCommand.java:196]: `id` is `toLowerCase`ed every time it's used, should probably lowercase it once at the start (and should use `CoreUtilites#toLowerCase`)

File: src/main/java/com/denizenscript/denizencore/scripts/commands/core/MongoCommand.java [hunk 15]
```diff
@@ -0,0 +1,521 @@
...
+                                   @ArgPrefixed @ArgDefaultNull @ArgName("update") MapTag oldData,
+                                   @ArgPrefixed @ArgDefaultNull @ArgName("new") MapTag newData,
+                                   @ArgName("upsert") boolean upsert,
+                                   @ArgPrefixed @ArgDefaultNull @ArgName("use_database") String newDatabase,
+                                   @ArgPrefixed @ArgDefaultNull @ArgName("use_collection") String newCollection) {
+        if (!CoreConfiguration.allowMongo) {
+            Debug.echoError(scriptEntry, "Mongo disabled by config!");
+            return;
+        }
+        if (id == null) {
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/denizencore/scripts/commands/core/MongoCommand.java:187]: check is redundant, it can't be null

File: src/main/java/com/denizenscript/denizencore/scripts/commands/core/MongoCommand.java [hunk 16]
```diff
@@ -0,0 +1,521 @@
...
+                        return;
+                    }
+                    DenizenCore.schedule(new AsyncSchedulable(new OneTimeSchedulable(() -> {
+                        MongoClient con = null;
+                        String conStr = uri.getValue();
+                        if (!conStr.startsWith("mongodb://") && !conStr.startsWith("mongodb+srv://")) {
+                            conStr = "mongodb://" + conStr;
+                        }
+                        if (CoreConfiguration.debugVerbose) {
+                            Debug.echoDebug(scriptEntry, "Connecting to Mongo server...");
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/denizencore/scripts/commands/core/MongoCommand.java:217]: Why is this here?

File: src/main/java/com/denizenscript/denizencore/scripts/commands/core/MongoCommand.java [hunk 17]
```diff
@@ -0,0 +1,521 @@
...
+    // - ~mongo id:name use_database:my_new_database
+    //
+    // @Usage
+    // Change Collections.
+    // - ~mongo id:name use_collection:my_new_collection
+    // -->
+
+    public static Map<String, MongoClient> connections = new HashMap<>();
+    public static Map<String, MongoDatabase> databases = new HashMap<>();
+    public static Map<String, MongoCollection<Document>> collections = new HashMap<>();
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/denizencore/scripts/commands/core/MongoCommand.java:145]: Rather than triple maps, there should be a single map and a sub-class containing all 3 values

File: src/main/java/com/denizenscript/denizencore/scripts/commands/core/MongoCommand.java [hunk 18]
```diff
@@ -0,0 +1,521 @@
...
+                                Debug.echoDebug(scriptEntry, "Connecting errored!");
+                                scriptEntry.setFinished(true);
+                            });
+                        }
+                    }, 0)));
+                }
+                // If disconnect is true, it means it is present.
+                else if (disconnect) {
+                    MongoClient con = connections.get(connectionId);
+                    if (con == null) {
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/denizencore/scripts/commands/core/MongoCommand.java:261]: Since this check is duplicated for everything other than the `connect` option, can just `return;` from the `connect` block and then handle the `con` grab outside the `if`

File: src/main/java/com/denizenscript/denizencore/scripts/commands/core/MongoCommand.java [hunk 19]
```diff
@@ -0,0 +1,521 @@
...
+                    }
+                    if (newData == null) {
+                        Debug.echoError(scriptEntry, "You must specify the new data to be updated!");
+                        scriptEntry.setFinished(true);
+                        return;
+                    }
+                    try {
+                        HashMap<String, Object> updateMap = (HashMap<String, Object>) CoreUtilities.objectTagToJavaForm(oldData, false, true);
+                        HashMap<String, Object> newMap = (HashMap<String, Object>) CoreUtilities.objectTagToJavaForm(newData, false, true);
+                        Debug.echoDebug(scriptEntry, "Updating data. . .");
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/denizencore/scripts/commands/core/MongoCommand.java:421]: shouldn't have spaces between the `...`

File: src/main/java/com/denizenscript/denizencore/scripts/commands/core/MongoCommand.java [hunk 20]
```diff
@@ -0,0 +1,442 @@
...
+                        Debug.echoDebug(scriptEntry, "Connecting to Mongo server...");
+                        MongoClient con = null;
+                        MongoDatabase db = null;
+                        MongoCollection<Document> col = null;
+                        try {
+                            con = MongoClients.create(conStr);
+                            db = con.getDatabase(database);
+                            col = db.getCollection(collection);
+                        }
+                        catch (final Exception e) {
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/denizencore/scripts/commands/core/MongoCommand.java:223]: this exception block never `return;`s, so it'll throw an error and then immediately follow with "connection did not error"

File: src/main/java/com/denizenscript/denizencore/scripts/commands/core/MongoCommand.java [hunk 21]
```diff
@@ -0,0 +1,442 @@
...
+                            DenizenCore.runOnMainThread(() -> {
+                                Debug.echoError(scriptEntry, "Mongo Exception: " + e.getMessage());
+                                scriptEntry.setFinished(true);
+                                if (CoreConfiguration.debugVerbose) {
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/denizencore/scripts/commands/core/MongoCommand.java:227]: Why is this verbose only?

File: src/main/java/com/denizenscript/denizencore/scripts/commands/core/MongoCommand.java [hunk 22]
```diff
@@ -0,0 +1,442 @@
...
+                                    Debug.echoError(scriptEntry, e);
+                                }
+                            });
+                        }
+                        if (CoreConfiguration.debugVerbose) {
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/denizencore/scripts/commands/core/MongoCommand.java:232]: Why is this verbose only?

File: src/main/java/com/denizenscript/denizencore/scripts/commands/core/MongoCommand.java [hunk 23]
```diff
@@ -0,0 +1,442 @@
...
+                            Debug.echoDebug(scriptEntry, "Connection did not error.");
+                        }
+                        final MongoClient conn = con;
+                        final MongoDatabase connDB = db;
+                        final MongoCollection<Document> coll = col;
+                        if (con != null) {
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/denizencore/scripts/commands/core/MongoCommand.java:238]: Why are there two separate error checks?

File: src/main/java/com/denizenscript/denizencore/scripts/commands/core/MongoCommand.java [hunk 24]
```diff
@@ -0,0 +1,442 @@
...
+        if (!CoreConfiguration.allowMongo) {
+            Debug.echoError(scriptEntry, "Mongo disabled by config!");
+            return;
+        }
+        String connectionId = CoreUtilities.toLowerCase(id);
+        Connection connection = mongoConnections.get(connectionId);
+        Runnable runnable = () -> {
+            try {
+                if (uri != null) {
+                    if (mongoConnections.containsKey(connectionId)) {
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/denizencore/scripts/commands/core/MongoCommand.java:194]: this could just be a `!= null` check since it's already grabbed earlier

File: src/main/java/com/denizenscript/denizencore/scripts/commands/core/MongoCommand.java [hunk 25]
```diff
@@ -0,0 +1,442 @@
...
+                        else {
+                            DenizenCore.runOnMainThread(() -> {
+                                Debug.echoDebug(scriptEntry, "Connecting errored!");
+                                scriptEntry.setFinished(true);
+                            });
+                        }
+                    }, 0)));
+                    return;
+                }
+                if (connection.connection == null) {
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/denizencore/scripts/commands/core/MongoCommand.java:258]: this will NPE as it's `connection` that will be null, not `connection.connection` - remember to test error cases, not just the good cases

File: src/main/java/com/denizenscript/denizencore/scripts/commands/core/MongoCommand.java [hunk 26]
```diff
@@ -0,0 +1,442 @@
...
+                    Debug.echoError(scriptEntry, "Not connected to server with ID: '" + id + "'!");
+                    scriptEntry.setFinished(true);
+                    return;
+                }
+                if (connection.database == null) {
+                    Debug.echoError(scriptEntry, "Not connected to database! Was it dropped?");
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/denizencore/scripts/commands/core/MongoCommand.java:264]: Not all input options require a database or collection. For example, `disconnect` doesn't require anything but the connection itself.

File: src/main/java/com/denizenscript/denizencore/scripts/commands/core/MongoCommand.java [hunk 27]
```diff
@@ -0,0 +1,425 @@
...
+                        });
+                    }, 0)));
+                    return;
+                }
+                if (connection == null) {
+                    Debug.echoError(scriptEntry, "There is no open connection with ID: '" + id + "'! Has it been disconnected?");
+                    scriptEntry.setFinished(true);
+                    return;
+                }
+                if (connection.connection == null) {
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/denizencore/scripts/commands/core/MongoCommand.java:252]: In what situation can this ever occur?
